winch/codegen/src/isa/x64/asm.rs

```diff
@@ -146,23 +167,42 @@ impl Assembler {
         &mut self.buffer
     }
 
+    /// Adds a constant to the constant pool and returns its address.
+    pub fn add_constant(&mut self, constant: &[u8]) -> Address {
+        let handle = self.pool.insert(constant.into());
+        Address::constant(handle)
+    }
+
     /// Return the emitted code.
     pub fn finalize(mut self) -> MachBufferFinalized<Final> {
         let stencil = self
             .buffer
-            .finish(&Default::default(), self.emit_state.ctrl_plane_mut());
+            .finish(&self.constants, self.emit_state.ctrl_plane_mut());
         stencil.apply_base_srcloc(Default::default())
     }
 
     fn emit(&mut self, inst: Inst) {
         inst.emit(&[], &mut self.buffer, &self.emit_info, &mut self.emit_state);
     }
 
-    fn to_synthetic_amode(addr: &Address) -> SyntheticAmode {
+    fn to_synthetic_amode(
+        addr: &Address,
+        pool: &mut ConstantPool,
+        constants: &mut VCodeConstants,
+        buffer: &mut MachBuffer<Inst>,
+    ) -> SyntheticAmode {
         match addr {
             Address::Offset { base, offset } => {
                 SyntheticAmode::real(Amode::imm_reg(*offset as i32, (*base).into()))
             }
+            Address::Const(c) => {
+                // XXX Can we do this in `add_constant`?
```

Comment:
I'm not sure exactly (I haven't paged in enough details of the constant pool) but thought I should call this `XXX` comment out -- can we resolve it either way (do the refactor or remove the comment and note why we're doing this)?

Reply:
Yeah, this comment is old and not applicable. I've removed it.